app/models/checklist_source_data_criteria.rb

```diff
@@ -9,27 +9,56 @@ class ChecklistSourceDataCriteria
   field :source_data_criteria, type: String # this is the name of the source_data_criteria
 
   field :completed, type: Boolean
+  field :recorded_result, type: String
   field :code, type: String
   field :attribute_code, type: String
 
-  validate :code_matches_valueset?
-  validate :attribute_code_matches_valueset?
+  field :code_complete, type: Boolean
+  field :attribute_complete, type: Boolean
+  field :result_complete, type: Boolean
+
+  def validate_criteria
+    result_completed?
+    attribute_code_matches_valueset?
+    code_matches_valueset?
+    complete?
+  end
+
+  def complete?
+    self.completed = if code.blank? && attribute_code.blank? && recorded_result.blank?
+                       nil
+                     else
+                       code_complete != false && attribute_complete != false && result_complete != false
+                     end
+  end
+
+  def result_completed?
```

Comment:
Is there a reason we're putting this `Boolean` value into a field, rather than just using this function?